kubelet: add resource allocation events

ffromani opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Add events for resource allocation, both for the happy path to give some explicit feedback about succesfull allocation, and improve the error path to add information to events to help the troubleshooting.

Notably, we can improve also the workload awareness, projecting somehow the information in the container (maybe environment variables), but this is deferred to a future change

[EDIT 20240214]
addressing possible drawbacks of events:

1. [short lived](https://www.cncf.io/blog/2023/03/13/how-to-use-kubernetes-events-for-effective-alerting-and-monitoring/) - but should not be a problem for this context: users can fetch the events after pod admission failure to gain more insights, or after the pod goes running.
2. event can be [throttled](https://kubernetes.slack.com/archives/C0EG7JC6T/p1708002244634309?thread_ts=1707930394.080309&cid=C0EG7JC6T):
  a.  the core/v1 API has a client-side throttle of 1000 concurrent events: https://github.com/openshift/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/record/event.go#L43
  b. the events/v1 API is merging event together and delaying their submission to reduce the pressure on the apiserver. Events will never be lost with this one, but their update on the apiserver side can be delayed by either 7 or 30 minutes

The conversation about [events](https://kubernetes.slack.com/archives/C0EG7JC6T/p1708001716829729?thread_ts=1707930394.080309&cid=C0EG7JC6T) further reinforces what emerged in the previous conversations: events are an improvement for fast-feedback, but to improve visibility/observability we will need to review and potentially enhance also metrics and logs. I'll update/file the issues 

#### Which issue(s) this PR fixes:
Part of the fix for #123037 

#### Special notes for your reviewer:
Please check the commit message(s) for full context and implementation rationale
We will need to review events and logs before to call 123037 done

#### Does this PR introduce a user-facing change?
```release-note
Add events to improve the observability of the resource allocation done by the kubelet
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
N/A
```

ffromani commented:
/sig node

ffromani commented:
/cc @rphillips @swatisehgal @Tal-or

ffromani commented on pkg/kubelet/cm/topologymanager/topology_manager.go:
```diff
@@ -92,6 +104,10 @@ type HintProvider interface {
+	// GetExclusiveResources returns a list of resource names whose instances were
+	// esclusively allocated to this container. This is used by the orchestrator to
+	// check if a container got exclusive (vs shared) resources or not.
+	GetExclusiveResources(pod *v1.Pod, container *v1.Container) []string
```
or we can just extend `Allocate` to report the same information

ffromani commented:
/hold

the PR is reviewable, all the concepts are there, but I'd like to improve the test coverage

swatisehgal commented:
/triage accepted

ffromani commented:
note: a number of unit test WILL fail because of the new errors being returned. I'm reviewing the failures and I'll fix them in the next uploads; a fair amount of tests could use some improvement because they were naively checking errors for equality. We can do better than this regardless of the changes being proposed here.

ffromani commented:
/retest

Tal-or commented on pkg/kubelet/cm/cpumanager/policy_static_test.go:
```diff
@@ -622,7 +623,7 @@ func runStaticPolicyTestCase(t *testing.T, testCase staticPolicyTest) {
 
 	container := &testCase.pod.Spec.Containers[0]
-	err := policy.Allocate(st, testCase.pod, container)
+	err := unwrapErr(policy.Allocate(st, testCase.pod, container))
```
What is the reason to return an `error` and then immediately unwrap it?

ffromani replied:
because the tests check errors for equality (e.g. they *not* use `errors.As` or `errors.Is`) and this is the quickest fix still being (IMO) (barely?) good enough.
We should probably clean the tests to use `errors`, but this is a larger endeavor. If reviewers feel otherwise though, fair enough, I can prepend this cleanup to this PR.

Tal-or replied:
Thank you for the explanation.
Yes we probably should wait from the other reviewers comments on that.

klueska replied:
I would prefer do the unwrap at the point of comparison, not at the return site.

Tal-or reviewed (commented):
Thank you for your PR. It definitely improve the observability and make it easier to track the allocation flow.

Left couple of comments, but looks good in general

ffromani commented on pkg/kubelet/cm/admission/events.go:
```diff
@@ -0,0 +1,91 @@
+	return e.Err
+}
+
+func MakeResourceAllocationError(reason, resource string, needed int, affinity fmt.Stringer, err error) ResourceAllocationError {
```
good point. Perhaps I should pass  a `resource.Quantity` here.

ffromani replied:
fixed the Memory Manager flow

ffromani commented on pkg/kubelet/cm/topologymanager/topology_manager.go:
```diff
@@ -43,10 +44,20 @@ const (
 func (e TopologyAffinityError) Error() string {
-	return "Resources cannot be allocated with Topology locality"
+	msg := "Resources cannot be allocated with Topology locality"
+	if e.ContainerName != "" {
```
Uhm, perhaps I should make `ContainerName` mandatory indeed. I'll re-evaluate the flows.

ffromani replied:
ContainerName can be None in the `scope=pod` flow

ffromani commented on pkg/kubelet/cm/topologymanager/topology_manager.go:
```diff
@@ -43,10 +44,20 @@ const (
 
 func (e TopologyAffinityError) Error() string {
-	return "Resources cannot be allocated with Topology locality"
+	msg := "Resources cannot be allocated with Topology locality"
```
In general, yes. In the specific context of generating the message for this error, I think the trivial string concatenation is the sweet spot because we are expected to do this rarely, and the code is very trivial to read.

ffromani commented:
will look like this:
```
[fromani@gojira numaresources-operator]$ kubectl describe pod numalign-pod-4pn2l
Name:             numalign-pod-4pn2l
Namespace:        default
Priority:         0
Service Account:  default
Node:             kind-worker/A.B.C.D
Start Time:       Mon, 12 Feb 2024 12:16:12 -0500
Labels:           <none>
Annotations:      <none>
Status:           Running
IP:               10.244.1.2
IPs:
  IP:  10.244.1.2
Containers:
  numalign:
    Container ID:  containerd://6f4fc9e932667ab2ea422c918bdfeadef1bc6a9be0026c54fc58f1bdf925ded8
    Image:         quay.io/fromani/numalign
    Image ID:      sha256:751cef88a0a95e0f5396c5fa6323a82a1cc3fd7c072e717890b0daba1474f2c9
    Port:          <none>
    Host Port:     <none>
    Command:
      /usr/local/bin/numalign
    State:          Running
      Started:      Mon, 12 Feb 2024 12:16:17 -0500
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     1
      memory:  256Mi
    Requests:
      cpu:     1
      memory:  256Mi
    Environment:
      NUMALIGN_SLEEP_HOURS:  127
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-sfs5s (ro)
Conditions:
  Type                        Status
  PodReadyToStartContainers   True 
  Initialized                 True 
  Ready                       True 
  ContainersReady             True 
  PodScheduled                True 
Volumes:
  kube-api-access-sfs5s:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Guaranteed
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type    Reason                     Age   From               Message
  ----    ------                     ----  ----               -------
  Normal  Scheduled                  15s   default-scheduler  Successfully assigned default/numalign-pod-4pn2l to kind-worker
  Normal  AllocatedAlignedResources  16s   kubelet            cpu: containers=1;memory: containers=1
  Normal  Pulling                    15s   kubelet            Pulling image "quay.io/fromani/numalign"
  Normal  Pulled                     11s   kubelet            Successfully pulled image "quay.io/fromani/numalign" in 4.378s (4.378s including waiting)
  Normal  Created                    11s   kubelet            Created container numalign
  Normal  Started                    11s   kubelet            Started container numalign
```

ffromani commented:
another update, with multiple containers:
```
[fromani@gojira deployer]$ K describe pod numalign-pod-n6qhh
Name:             numalign-pod-n6qhh
Namespace:        default
Priority:         0
Service Account:  default
Node:             kind-worker/A.B.C.D
Start Time:       Tue, 13 Feb 2024 06:27:52 -0500
Labels:           <none>
Annotations:      <none>
Status:           Running
IP:               10.244.1.4
IPs:
  IP:  10.244.1.4
Containers:
  numalign-0:
    Container ID:  containerd://5d72e0c452578d0744075f225f8a803c02ef69eb65706dd785b0bb9ffc2c48d5
    Image:         quay.io/fromani/numalign
    Image ID:      sha256:751cef88a0a95e0f5396c5fa6323a82a1cc3fd7c072e717890b0daba1474f2c9
    Port:          <none>
    Host Port:     <none>
    Command:
      /usr/local/bin/numalign
    State:          Running
      Started:      Tue, 13 Feb 2024 06:27:52 -0500
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     1
      memory:  256Mi
    Requests:
      cpu:     1
      memory:  256Mi
    Environment:
      NUMALIGN_SLEEP_HOURS:  127
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-vg4fz (ro)
  numalign-1:
    Container ID:  containerd://ad49c2df7a5a5f1d5dff36080d30827ab9dfbbd22bedabd8a94558bc1d69d89d
    Image:         quay.io/fromani/numalign
    Image ID:      sha256:751cef88a0a95e0f5396c5fa6323a82a1cc3fd7c072e717890b0daba1474f2c9
    Port:          <none>
    Host Port:     <none>
    Command:
      /usr/local/bin/numalign
    State:          Running
      Started:      Tue, 13 Feb 2024 06:27:52 -0500
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     1
      memory:  256Mi
    Requests:
      cpu:     1
      memory:  256Mi
    Environment:
      NUMALIGN_SLEEP_HOURS:  127
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-vg4fz (ro)
  numalign-2:
    Container ID:  containerd://435eabec828622c430b2caf0dc2a180e667bec8a829b44695cddf6147a5980d5
    Image:         quay.io/fromani/numalign
    Image ID:      sha256:751cef88a0a95e0f5396c5fa6323a82a1cc3fd7c072e717890b0daba1474f2c9
    Port:          <none>
    Host Port:     <none>
    Command:
      /usr/local/bin/numalign
    State:          Running
      Started:      Tue, 13 Feb 2024 06:27:53 -0500
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     1
      memory:  256Mi
    Requests:
      cpu:     1
      memory:  256Mi
    Environment:
      NUMALIGN_SLEEP_HOURS:  127
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-vg4fz (ro)
  numalign-3:
    Container ID:  containerd://fdbb504e0abe3da750eb7db6192411cc9a391284c2a1037df1eefee622e9a960
    Image:         quay.io/fromani/numalign
    Image ID:      sha256:751cef88a0a95e0f5396c5fa6323a82a1cc3fd7c072e717890b0daba1474f2c9
    Port:          <none>
    Host Port:     <none>
    Command:
      /usr/local/bin/numalign
    State:          Running
      Started:      Tue, 13 Feb 2024 06:27:53 -0500
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     1
      memory:  256Mi
    Requests:
      cpu:     1
      memory:  256Mi
    Environment:
      NUMALIGN_SLEEP_HOURS:  127
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-vg4fz (ro)
  numalign-4:
    Container ID:  containerd://dc0d2f4e5c75323752bdeba451444050b089a6350ba87c6147118c60b80710e9
    Image:         quay.io/fromani/numalign
    Image ID:      sha256:751cef88a0a95e0f5396c5fa6323a82a1cc3fd7c072e717890b0daba1474f2c9
    Port:          <none>
    Host Port:     <none>
    Command:
      /usr/local/bin/numalign
    State:          Running
      Started:      Tue, 13 Feb 2024 06:27:53 -0500
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     1
      memory:  256Mi
    Requests:
      cpu:     1
      memory:  256Mi
    Environment:
      NUMALIGN_SLEEP_HOURS:  127
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-vg4fz (ro)
  numalign-5:
    Container ID:  containerd://db441aa80328b928f78bf1f5452347b91b6878d3bb679bea88fa16a9ce585a98
    Image:         quay.io/fromani/numalign
    Image ID:      sha256:751cef88a0a95e0f5396c5fa6323a82a1cc3fd7c072e717890b0daba1474f2c9
    Port:          <none>
    Host Port:     <none>
    Command:
      /usr/local/bin/numalign
    State:          Running
      Started:      Tue, 13 Feb 2024 06:27:53 -0500
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     1
      memory:  256Mi
    Requests:
      cpu:     1
      memory:  256Mi
    Environment:
      NUMALIGN_SLEEP_HOURS:  127
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-vg4fz (ro)
  numalign-6:
    Container ID:  containerd://2f7391de79656386619ccab4f5e377718f24bd4fa75d8651c8f6944b9e95d588
    Image:         quay.io/fromani/numalign
    Image ID:      sha256:751cef88a0a95e0f5396c5fa6323a82a1cc3fd7c072e717890b0daba1474f2c9
    Port:          <none>
    Host Port:     <none>
    Command:
      /usr/local/bin/numalign
    State:          Running
      Started:      Tue, 13 Feb 2024 06:27:53 -0500
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     1
      memory:  256Mi
    Requests:
      cpu:     1
      memory:  256Mi
    Environment:
      NUMALIGN_SLEEP_HOURS:  127
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-vg4fz (ro)
  numalign-7:
    Container ID:  containerd://00e29e1b017808f658b0aa68c2814e4ea1f790f4e9a6a65a9398e40a90bee2f9
    Image:         quay.io/fromani/numalign
    Image ID:      sha256:751cef88a0a95e0f5396c5fa6323a82a1cc3fd7c072e717890b0daba1474f2c9
    Port:          <none>
    Host Port:     <none>
    Command:
      /usr/local/bin/numalign
    State:          Running
      Started:      Tue, 13 Feb 2024 06:27:53 -0500
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     1
      memory:  256Mi
    Requests:
      cpu:     1
      memory:  256Mi
    Environment:
      NUMALIGN_SLEEP_HOURS:  127
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-vg4fz (ro)
  numalign-8:
    Container ID:  containerd://0dc6dccdf27ffa0be2b5fcbc80e994614df0ad0b80454678aaea26419fc0d6d8
    Image:         quay.io/fromani/numalign
    Image ID:      sha256:751cef88a0a95e0f5396c5fa6323a82a1cc3fd7c072e717890b0daba1474f2c9
    Port:          <none>
    Host Port:     <none>
    Command:
      /usr/local/bin/numalign
    State:          Running
      Started:      Tue, 13 Feb 2024 06:27:53 -0500
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     1
      memory:  256Mi
    Requests:
      cpu:     1
      memory:  256Mi
    Environment:
      NUMALIGN_SLEEP_HOURS:  127
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-vg4fz (ro)
  numalign-9:
    Container ID:  containerd://f7fa1283633c85900c1f30d636b6db991abe8ad872603f68678610d6c4f6370f
    Image:         quay.io/fromani/numalign
    Image ID:      sha256:751cef88a0a95e0f5396c5fa6323a82a1cc3fd7c072e717890b0daba1474f2c9
    Port:          <none>
    Host Port:     <none>
    Command:
      /usr/local/bin/numalign
    State:          Running
      Started:      Tue, 13 Feb 2024 06:27:53 -0500
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     1
      memory:  256Mi
    Requests:
      cpu:     1
      memory:  256Mi
    Environment:
      NUMALIGN_SLEEP_HOURS:  127
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-vg4fz (ro)
Conditions:
  Type                        Status
  PodReadyToStartContainers   True 
  Initialized                 True 
  Ready                       True 
  ContainersReady             True 
  PodScheduled                True 
Volumes:
  kube-api-access-vg4fz:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Guaranteed
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type    Reason     Age   From               Message
  ----    ------     ----  ----               -------
  Normal  Scheduled  119s  default-scheduler  Successfully assigned default/numalign-pod-n6qhh to kind-worker
  Normal  Allocated  119s  kubelet            cpu: containers=10; memory: containers=10
  Normal  Pulled     119s  kubelet            Container image "quay.io/fromani/numalign" already present on machine
  Normal  Created    119s  kubelet            Created container numalign-0
  Normal  Started    119s  kubelet            Started container numalign-0
  Normal  Pulled     119s  kubelet            Container image "quay.io/fromani/numalign" already present on machine
  Normal  Created    119s  kubelet            Created container numalign-1
  Normal  Started    119s  kubelet            Started container numalign-1
  Normal  Pulled     119s  kubelet            Container image "quay.io/fromani/numalign" already present on machine
  Normal  Created    119s  kubelet            Created container numalign-2
  Normal  Started    118s  kubelet            Started container numalign-2
  Normal  Pulled     118s  kubelet            Container image "quay.io/fromani/numalign" already present on machine
  Normal  Created    118s  kubelet            Created container numalign-3
  Normal  Started    118s  kubelet            Started container numalign-3
  Normal  Pulled     118s  kubelet            Container image "quay.io/fromani/numalign" already present on machine
  Normal  Created    118s  kubelet            Created container numalign-4
  Normal  Started    118s  kubelet            Started container numalign-4
  Normal  Pulled     118s  kubelet            Container image "quay.io/fromani/numalign" already present on machine
  Normal  Created    118s  kubelet            Created container numalign-5
  Normal  Started    118s  kubelet            Started container numalign-5
  Normal  Pulled     118s  kubelet            Container image "quay.io/fromani/numalign" already present on machine
  Normal  Created    118s  kubelet            Created container numalign-6
  Normal  Started    118s  kubelet            Started container numalign-6
  Normal  Pulled     118s  kubelet            Container image "quay.io/fromani/numalign" already present on machine
  Normal  Created    118s  kubelet            Created container numalign-7
  Normal  Started    118s  kubelet            Started container numalign-7
```

ffromani commented:
a (trivial) failure: more example will follow:
```
Name:             numalign-pod-b7w46
Namespace:        default
Priority:         0
Service Account:  default
Node:             kind-worker/
Start Time:       Tue, 13 Feb 2024 16:50:07 +0100
Labels:           <none>
Annotations:      <none>
Status:           Failed
Reason:           TopologyAffinityError
Message:          Pod was rejected: Resources cannot be allocated with Topology locality container=numalign hint=[affinity=N/A preferred=false]
IP:               
IPs:              <none>
Containers:
  numalign:
    Image:      quay.io/fromani/numalign
    Port:       <none>
    Host Port:  <none>
    Command:
      /usr/local/bin/numalign
    Limits:
      cpu:     16
      memory:  4000Mi
    Requests:
      cpu:     16
      memory:  4000Mi
    Environment:
      NUMALIGN_SLEEP_HOURS:  127
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-x9n5v (ro)
Volumes:
  kube-api-access-x9n5v:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Guaranteed
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason                 Age   From               Message
  ----     ------                 ----  ----               -------
  Normal   Scheduled              16s   default-scheduler  Successfully assigned default/numalign-pod-b7w46 to kind-worker
  Warning  TopologyAffinityError  16s   kubelet            Resources cannot be allocated with Topology locality container=numalign hint=[affinity=N/A preferred=false]
```

ffromani commented:
/hold cancel

klueska commented on pkg/kubelet/cm/devicemanager/types.go:
```diff
@@ -81,6 +81,10 @@ type Manager interface {
 
 	// UpdateAllocatedDevices frees any Devices that are bound to terminated pods.
 	UpdateAllocatedDevices()
+
+	// GetExclusiveResources returns a list of resource names whose instances were esclusively
+	// allocated to this container.
+	GetExclusiveResources(pod *v1.Pod, container *v1.Container) []string
```
For symmetry with the other functions in this interface, should this be called `GetAllocatedDevices()`? Also, how is this different than `GetDevices()` in terms of the underlying information that it provides?

ffromani replied:
addressed below

klueska commented on pkg/kubelet/cm/memorymanager/memory_manager.go:
```diff
@@ -91,6 +91,10 @@ type Manager interface {
 
 	// GetMemory returns the memory allocated by a container from NUMA nodes
 	GetMemory(podUID, containerName string) []state.Block
+
+	// GetExclusiveResources returns a list of resource names whose instances were esclusively
+	// allocated to this container.
+	GetExclusiveResources(pod *v1.Pod, container *v1.Container) []string
```
I see, it's the same name here too. It should probably be its own interface then, rather than embedded in both the DeviceManager and MemoryManager interfaces.

ffromani replied:
answering myself, because at least `GetDeviceRunContainerOptions` in device manager, extending 'Allocate' may be trickier than anticipated. It _seems_ that a getter is a simpler option overall.

klueska commented on pkg/kubelet/cm/memorymanager/policy_static.go:
```diff
@@ -327,7 +340,7 @@ func getPodRequestedResources(pod *v1.Pod) (map[v1.ResourceName]uint64, error) {
 	// Total resources requested by restartable init containers.
 	reqRsrcsByRestartableInitCtrs := make(map[v1.ResourceName]uint64)
 	for _, ctr := range pod.Spec.InitContainers {
-		reqRsrcs, err := getRequestedResources(pod, &ctr)
+		reqRsrcs, _, err := getRequestedResources(pod, &ctr)
```
It's odd to me to have a function that "most of the time" has one of its return values ignored. It feels like there should be a better abstraction here.

ffromani replied:
good point. I need the extra return value only for better error message, to express memory in terms of human friendl units (e.g. `1000Mi`) rather than in bytes. I'd think about something nicer.

ffromani replied:
I _think_ I found a better way given how we want to use the data

ffromani commented:
/hold

not all the comments have been addressed yet

kannon92 commented on test/e2e_node/cpu_manager_metrics_test.go:
```diff
@@ -189,7 +189,7 @@ func makeGuaranteedCPUExclusiveSleeperPod(name string, cpus int) *v1.Pod {
 						},
 					},
-					Command: []string{"sh", "-c", "sleep", "1d"},
+					Command: []string{"/bin/sleep", "86400"},
```
I prefer the 1d versus 86400. Any reason for the change?

ffromani replied:
yes, apparently the image changed and is not accepting the old value. I'll handle this on a separate PR, though.

ffromani commented:
need to check about event throttling. Depending on how it goes, we may lose the signal

kannon92 reviewed (commented):
As you mentioned, events can be throttled so you may miss these events on busy clusters.

ffromani commented:
another note to self: I' need to try harder to split this huge change in bits easier to review

ffromani commented:
> As you mentioned, events can be throttled so you may miss these events on busy clusters.

Thanks Kevin. I think clarifying the drawbacks of events is important before to move forward. I'll keep the hold until this is done. ATM I'm still inclined towards keeping this work, because the only alternatives to improve UX besides events seems to be metrics (already in the pipeline, PR TBD), more/better logs (https://github.com/kubernetes/kubernetes/pull/120569) or changes to pod.Status, which is a much larger endeavour.

ffromani commented:
/retest

aojea commented:
why don't expose metrics instead of events?
events are very noisy, short lived and unreliable, also adds a lot of pressure on the control plane since they are stored in etcd and a large number of events grows the db ... 

I suggest metrics and logs messages are the best option to provide observability , events only on very specific and clear cases so you avoid having noise, otherwise we are building a kind of centralized syslog with one hour retention, that is not solving most of the problems

ffromani commented:
> why don't expose metrics instead of events? 

We are already exposing metrics, but this is not helping to get direct feedback about how resource allocations are going
(see: https://github.com/kubernetes/kubernetes/pull/112855 https://github.com/kubernetes/kubernetes/pull/115590 https://github.com/kubernetes/kubernetes/pull/115137 https://github.com/kubernetes/kubernetes/pull/121778)

> events are very noisy, short lived and unreliable, also adds a lot of pressure on the control plane since they are stored in etcd and a large number of events grows the db ...

For that, I'm intentionally setting a hard upper bound to the number of extra events: up to 1 (one) in the happy path, up to 2 (two) in the error path

> I suggest metrics and logs messages are the best option to provide observability , events only on very specific and clear cases so you avoid having noise, otherwise we are building a kind of centralized syslog with one hour retention, that is not solving most of the problems

This is what we have right now, but from direct experience is not ideal. And it was also highlighted as graduation blocker and general concern during PRR (with I agree with): https://kubernetes.slack.com/archives/CPNHUMN74/p1696518756979339?thread_ts=1696357853.594279&cid=CPNHUMN74

ffromani committed:
node: memorymanager: add more logs

add more logs (at high verb levels) to help troubleshoot
allocations.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
node: topomgr: extend TopologyAffinityError

Add more information to TopologyAffinityError; this
information will be used later to produce events,
and possibly to make logs richer, in order to make
the operation of TopologyManager more visible.

The new information is optional and will be filled
(and used) in followup PRs.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
node: e2e: tests for resource allocation events

Add tests to exercise the new allocation events.
Because it's hard to predict how the message would
look like - this will also depend on the available
hardware on which the test run (e.g. CI), we
limit the scope of the test.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
node: kubelet: send events to convey allocation outcome

The kubelet didn't provide explicit feedback about esclusive
resource allocation.  The implicit contract was that if the
pod pass admission, then the required resources were allocated
respecting the configuration, for example cpu pinning or NUMA
node alignment.

Furthermore, the kubelet provided no explicit feedback about
failed admission because of failed resource allocation, with
feedback and reason only given in the kubelet logs (but oftentimes
only at verbosiness level >= 4).

In order to improve the observability of the resource allocation,
we add kubernetes events to notify users (humans or programs)
and give more explicit feedback about how the allocation went and
why.

The implementation of the events poses few questions.

First and foremost, which component should emit the events?
Since its graduation to GA, the topology manager is the resource
allocation orchestrator even if it is functionally disabled using
"none" policy (no alignment at all is attempted).
Since resource allocation is done here anyway, it seems natural
to extend the topology manager codebase a little to add events.

Next, another question is how many events should we emit?
The resources are attached to the containers, but the allocation
currently happens at pod admission time.
Adding a cumulative event per-pod saves resources - we have a fixed
upper bound about the number of events we add, but leaks the
implementation detail that resource allocation is done at pod admission
time, which may change in the future.
However there's no clear plan yet, and this detail is already leaking
(admission failed because TopologyAffinityError).

In the error path, likewise we consolidate the event creation
at topology manager (orchestrator) level to set a fixed upper bound
to the amount of new events we generate.

Improving the workload awareness about resource being allocated
(e.g. projecting somehow the allocated resources in the containers)
will be addressed in future changes.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
node: topomgr: generalize AddHintProvider

Generalize the AddHintProvider method to register
providers, in order to enable a followup patch
to register more providers.
We want more providers to report allocation status,
so we can have better logs or generate kubernetes events,
to improve the visibility of the operations.

No intended change in behavior.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
node: topomgr: add AllocationInfoProvider

In order to expose more information (e.g. events)
about how the resource allocation went, we need a way
to know about the current allocation for a container.

We thus add a new interface to fetch the current data,
to be used at the end of the allocation path to report
the status.

Another option would have be to extend the Allocate()
function, but at the moment we prefer to not overextend
that interface and keep it targeted.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
the PR is reviewable again. Reviewers please be aware I'm still looking for a better approach regarding error unwrapping in unit tests.

ffromani commented:
> Reviewers please be aware I'm still looking for a better approach regarding error unwrapping in unit tests.

Improved in the very last upload

klueska commented on pkg/kubelet/cm/memorymanager/policy_static_test.go:
```diff
@@ -1561,9 +1566,10 @@ func TestStaticPolicyAllocate(t *testing.T) {
-			topologyHint:  &topologymanager.TopologyHint{NUMANodeAffinity: newNUMAAffinity(1), Preferred: true},
+			pod:               getPod("pod2", "container2", requirementsGuaranteed),
+			expectedError:     fmt.Errorf("[memorymanager] failed to find the extended preferred hint"),
+			shouldUnwrapError: true,
```
is this really necessary? I don't understand when you would ever *not* unwrap it.

Meaning, can't you just do the following:
```
			err = p.Allocate(s, testCase.pod, &testCase.pod.Spec.Containers[0])
			if err ! nil {
				err = errors.Unwrap(err)
			}
```

ffromani replied:
nope, this is is my attempt to avoid to always unwrap the error, but if we prefer to do the unwrapping unconditionally, I'm all for it - I think in some cases is the most sane solution. Fixing.

klueska commented on pkg/kubelet/cm/topologymanager/topology_manager.go:
```diff
@@ -57,9 +68,10 @@ func (e TopologyAffinityError) Type() string {
-	// AddHintProvider adds a hint provider to manager to indicate the hint provider
-	// wants to be consulted with when making topology hints
-	AddHintProvider(HintProvider)
+	// RegisterProvider adds a hint provider to manager to indicate the hint provider
+	// wants to be consulted with when making topology hints; the given AllocationInfoProvider
+	// is authoritative about the current resource allocation.
+	RegisterProvider(h HintProvider, a AllocationInfoProvider)
```
It's weird to me to have this function take two parameters if its the same underlying object always implementing both interfaces) . Either we should come up with a single, new abstraction that makes sense as a Hint/AllocationInfo provider together (because they always *must* be implemented by the same object), or two separate calls (in case some hint provider doesn't actually want to provide allocation info (though I don't see when that would ever be the case).

ffromani replied:
Thanks, very valid point. I think the best option is to have a new abstraction for the very reason you mentioned. I'll think about something. Naming will be fun, I'm exploring on the lines of `ResourceAllocator`

ffromani committed:
node: test: mockHintProvider -> fakeResourceAllocator

trivial rename to better reflect the new code purpose

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
/retest-required

ffromani commented:
/retest-required

ffromani commented:
/retest-required

ffromani commented:
/retest-required

ffromani commented:
/retest

failing to create a pod to run the linter is hardly this PR's fault

swatisehgal commented on pkg/kubelet/cm/memorymanager/policy_static.go:
```diff
@@ -140,12 +143,18 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
 			return fmt.Errorf("[memorymanager] failed to find the default preferred hint")
 		}
+
+		klog.V(4).InfoS("Topology Hint recomputed", "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "previous", bestHint, "recomputed", defaultHint)
```
I wonder if this should be log level 3 as well? Seems important enough to be at least log level 3 to me.

ffromani replied:
I think you're right, fixing.

swatisehgal commented on pkg/kubelet/cm/memorymanager/policy_static.go:
```diff
@@ -154,6 +163,8 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
 			return fmt.Errorf("[memorymanager] failed to find the extended preferred hint")
 		}
+
+		klog.V(4).InfoS("Topology Hint extended", "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "previous", bestHint, "extended", extendedHint)
```
Same as above.

ffromani replied:
I think you're right, fixing.

swatisehgal commented on pkg/kubelet/cm/topologymanager/topology_manager.go:
```diff
@@ -108,6 +109,7 @@ type HintProvider interface {
 // which can allocate resources, provide allocation hints.
 type ResourceAllocator interface {
 	HintProvider
+	AllocationInfoProvider
```
Rather than creating  a new interface (ResourceAllocator) why don't we embed `AllocationInfoProvider` directly in HintProvider interface or add `GetExclusiveResources` to the HintProvider interface?

ffromani replied:
This stems from my interpretation of a previous comment from @klueska . We can shoehorn the AllocationInfoProvider into the HintProvider, but feels like... forced. OTOH the two interfaces are related, so the option I like more atm is to create a new abstraction which includes both.

I'm open to suggestions though.

swatisehgal replied:
Thanks for the explanation, I was mostly curious and wanted to understand the thought process here.

swatisehgal reviewed (commented):
Looks good, left some minor comments/questions.

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -93,6 +93,11 @@ type Manager interface {
 	GetCPUAffinity(podUID, containerName string) cpuset.CPUSet
+
+	// GetExclusiveResources returns a list of resource names whose instances were
+	// esclusively allocated to this container. This is used by the orchestrator to
```
thanks, fixed (italianism :) )

ffromani commented:
/test pull-crio-cgroupv1-node-e2e-resource-managers
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/retest-required

ffromani commented:
/retest-required
/test pull-crio-cgroupv1-node-e2e-resource-managers
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

swatisehgal reviewed (commented):
I think Code-wise we are good, but there is some room for improvement in e2e tests.

ffromani commented on test/e2e_node/resource_allocation_events_test.go:
```diff
@@ -0,0 +1,158 @@
+			ginkgo.By("Creating the test pod which will be admitted")
+			testPod = e2epod.NewPodClient(f).CreateSync(ctx, makeGuaranteedCPUExclusiveSleeperPod("pin-alloc-good", threadsPerCore))
+
+			events, err := getEventsForPod(ctx, f.ClientSet, testPod.Namespace, testPod.Name)
```
waiting for the event with a timeout looks like a nice idea, thanks!

ffromani commented on test/e2e_node/resource_allocation_events_test.go:
```diff
@@ -0,0 +1,158 @@
+		}
+	}
+	return nil, false
+}
```
done, thanks!

ffromani commented:
/hold

no longer happy with the event on the happy path

ffromani commented:
/test pull-kubernetes-unit

ffromani commented:
/test pull-crio-cgroupv1-node-e2e-resource-managers
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/retest-required

ffromani commented:
/remove-lifecycle stale

ffromani commented:
after some time and after more considerations, I'm now a tad more like the position @aojea had previously (https://github.com/kubernetes/kubernetes/pull/123092#issuecomment-1945835043)

We *will* add metrics, more and more, but we need also a different form of feedback. Events are indeed not the best vehicle, especially at scale. But what other options do we have? Adding to pod status may be an option, but: omitting the fact is a much bigger change, we will be bubbling up details which I'm not sure should be exposed so directly to the user.

What I'm trying to say is that once again I acknowledge that events are not the best choice, yet it seems the best (ore less worse?) option available atm.

Going to rebase and refresh the PR ASAP.

ffromani commented:
what if instead of events we convey the same information (or as much as possible) in `pod.Status.Conditions`?

ffromani commented:
> what if instead of events we convey the same information (or as much as possible) in `pod.Status.Conditions`?

I'll propose a KEP for 1.34 and start sketching out a design doc when time allows. I'm more and more convinced that events are an imperfect solution to a real problem, so I'm willing to explore this option.
